`Pair Programming EDA II Nulos y valores extremos`

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import sidetable


In [2]:
df=pd.read_csv('../Datos/attacks_limpieza_I.csv')
df.head(2)

case_number  year        type country        area  \
0    2018.06.25  2018     Boating     usa  California   
1  2018.06.03.a  2018  Unprovoked  brazil  Pernambuco   

                      location  activity                    name age  \
0  Oceanside, San Diego County  Paddling             Julie Wolfe  57   
1        Piedade Beach, Recife  Swimming  Jose Ernesto da Silva   18   

             time      species        siglo species_  \
0           18h00  White shark  siglo - XXI      NaN   
1  Late afternoon  Tiger shark  siglo - XXI      NaN   

                                              injury  mes fatal_bien  \
0  No injury to occupant, outrigger canoe and pad...  Jun          N   
1                                              FATAL  Jun          Y   

  genero_bien categorias_tiburones  edades  
0           F                White    57.0  
1           M                Tiger    18.0



Objetivos
Usad la librería sidetable para:
1. Explorar los valores nulos.


In [4]:
df1= df.stb.missing()
df1

missing  total    percent
species_                 1628   1672  97.368421
species                   170   1672  10.167464
siglo                     170   1672  10.167464
time                      160   1672   9.569378
edades                    158   1672   9.449761
age                       154   1672   9.210526
mes                       154   1672   9.210526
fatal_bien                 99   1672   5.921053
location                   51   1672   3.050239
area                       46   1672   2.751196
activity                   31   1672   1.854067
country                    10   1672   0.598086
name                        9   1672   0.538278
injury                      8   1672   0.478469
year                        0   1672   0.000000
type                        0   1672   0.000000
genero_bien                 0   1672   0.000000
categorias_tiburones        0   1672   0.000000
case_number                 0   1672   0.000000

In [5]:
#comprobacion con la columna 'species'
df1.iloc[1]

missing     170.000000
total      1672.000000
percent      10.167464
Name: species, dtype: float64

In [6]:
df.sample(5)

case_number  year        type       country                   area  \
1199    1967.12.30  1967  Unprovoked  south africa  Western Cape Province   
32      2017.08.03  2017    Provoked           usa          Massachusetts   
644     2003.07.04  2003  Unprovoked       bahamas          Abaco Islands   
406     2008.12.10  2008    Provoked  south africa  Western Cape Province   
858   1996.09.06.a  1996  Unprovoked           usa                Florida   

                                           location            activity  \
1199                                     Mossel Bay             Surfing   
32    Menauhant Beach, Falmouth,\nBarnstable County             Fishing   
644                   10 miles west of Walker’s Cay        Spearfishing   
406                                 Plettenberg Bay             Fishing   
858                New Smyrna Beach, Volusia County  Surfing / Swimming   

                name age   time                          species        siglo  \
1199   Brian Pearson  22  14h00  1.5 m to 1.8 m [5' to 6'] shark     siglo-XX   
32              male  34  23h30                         4' shark  siglo - XXI   
644   Benjamin Brown  39  14h45           2.1 m  [7'] bull shark  siglo - XXI   
406      Luke Parker  15  20h00            Raggedtooth shark, 2m  siglo - XXI   
858     John Perkins  19  12h15                  "a young shark"     siglo-XX   

     species_                                             injury  mes  \
1199      NaN                          Right ankle & foot bitten  Dec   
32        NaN      Foot bitten by landed shark PROVOKED INCIDENT  Aug   
644       NaN                                   Left calf bitten  Jul   
406       NaN  Lacerations to knees, thigh and hip by hooked ...  Dec   
858       NaN  Small puncture wounds and lacerations on right...  Sep   

     fatal_bien genero_bien categorias_tiburones  edades  
1199          N           M          Unspecified    22.0  
32            N           M          Unspecified    34.0  
644           N           M                 Bull    39.0  
406           N           M          Unspecified    15.0  
858           N           M          Unspecified    19.0

Analizando el dataframe y las columnas nos dimos cuenta que hay dos columnas 'species' que contiene la misma informacion, pero una de ellas tiene 97% nulos y la otra solamente un aprox 10%, por lo que podriamos concluir que no es necesario conservar la primera columna.´
Las demas columnas no tienen un numero relevante de nulos, por lo tanto los vamos a conservar

2. Explorad la columna de country y reportad los resultados solo para aquellas categorías que incluyan el 70% de los datos.


In [7]:
df.stb.freq(['country'], thresh= 70)

/Users/feliciano/opt/anaconda3/lib/python3.9/site-packages/sidetable/sidetable.py:151: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  results = results[results[other_label] == False].append(


country  count    percent  cumulative_count  cumulative_percent
0        usa    741  44.584838               741           44.584838
1  australia    343  20.637786              1084           65.222623
2     others    578  34.777377              1662          100.000000

En este Df vemos que el 70% de nuestros datos con respeto a los ataques de tiburones corresponden a Usa y Australia

3. Explorad la columna country en relación al sexo, filtrando de nuevo por aquellas categorías que incluyan el 70% de los datos.


In [8]:
df.stb.freq(['country', 'genero_bien'], thresh= 70)

/Users/feliciano/opt/anaconda3/lib/python3.9/site-packages/sidetable/sidetable.py:151: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  results = results[results[other_label] == False].append(


country genero_bien  count    percent  cumulative_count  \
0           usa           M    604  36.341757               604   
1     australia           M    305  18.351384               909   
2  south africa           M    222  13.357401              1131   
3        others      others    531  31.949458              1662   

   cumulative_percent  
0           36.341757  
1           54.693141  
2           68.050542  
3          100.000000

Con respeto al genero el 70% de nuestros datos reflejan que el genero masculino es el que mas sufre ataques en los paises Usa, Australia y South Africa.

4. Haced los mismo para la columna activity en relación a la columna fatal. En este caso filtrad y mostrad los datos que incluyan el 60% de los datos.


In [9]:
df.stb.freq(['activity', 'fatal_bien'], thresh= 60)

/Users/feliciano/opt/anaconda3/lib/python3.9/site-packages/sidetable/sidetable.py:151: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  results = results[results[other_label] == False].append(


activity fatal_bien  count    percent  cumulative_count  \
0         Surfing          N    380  24.531956               380   
1        Swimming          N    139   8.973531               519   
2    Spearfishing          N     85   5.487411               604   
3        Swimming          Y     79   5.100065               683   
4         Fishing          N     50   3.227889               733   
5          Wading          N     43   2.775985               776   
6        Standing          N     30   1.936733               806   
7      Snorkeling          N     29   1.872176               835   
8         Surfing          Y     26   1.678502               861   
9   Body boarding          N     24   1.549387               885   
10         Diving          N     19   1.226598               904   
11   Scuba diving          N     17   1.097482               921   
12         others     others    628  40.542285              1549   

    cumulative_percent  
0            24.531956  
1            33.505487  
2            38.992899  
3            44.092963  
4            47.320852  
5            50.096837  
6            52.033570  
7            53.905746  
8            55.584248  
9            57.133635  
10           58.360232  
11           59.457715  
12          100.000000

- En general vemos que la mayoria de los ataques no son mortales.
- Las tres principales actividades donde mas ataques no fatales hay son surfing, swimming y spearfishing. 
- Esta tabla nos ayuda ver que las actividades con mayor numero de ataques es surfing con un 24.5% siendo estos no fatales, mientras que la actividad con mayor ataques fatales es nadar, pero solo con un 5% fatales

5. Por último, realizad este análisis para la columna fecha (la que creamos en el último ejercicio de pair programming de Pandas) en relación a la columna fatal. Mostrad solo los resultados para las categorías que incluyan el 60% de los datos.

In [10]:
df.stb.freq(['year', 'fatal_bien'], thresh=60)

/Users/feliciano/opt/anaconda3/lib/python3.9/site-packages/sidetable/sidetable.py:151: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  results = results[results[other_label] == False].append(


year fatal_bien  count    percent  cumulative_count  cumulative_percent
0      0.0          N     64   4.068659                64            4.068659
1   2015.0          N     59   3.750795               123            7.819453
2      0.0          Y     59   3.750795               182           11.570248
3   2007.0          N     50   3.178640               232           14.748887
4   2002.0          N     49   3.115067               281           17.863954
5   2012.0          N     47   2.987921               328           20.851875
6   2013.0          N     41   2.606484               369           23.458360
7   2006.0          N     36   2.288620               405           25.746980
8   2005.0          N     36   2.288620               441           28.035601
9   2003.0          N     36   2.288620               477           30.324221
10  2009.0          N     35   2.225048               512           32.549269
11  2010.0          N     32   2.034329               544           34.583598
12  2004.0          N     32   2.034329               576           36.617928
13  2016.0          N     31   1.970757               607           38.588684
14  2008.0          N     30   1.907184               637           40.495868
15  2014.0          N     28   1.780038               665           42.275906
16  2011.0          N     26   1.652893               691           43.928798
17  2001.0          N     26   1.652893               717           45.581691
18  2000.0          N     26   1.652893               743           47.234584
19  2017.0          N     23   1.462174               766           48.696758
20  1999.0          N     23   1.462174               789           50.158932
21  1961.0          N     20   1.271456               809           51.430388
22  1995.0          N     19   1.207883               828           52.638271
23  1996.0          N     18   1.144310               846           53.782581
24  1992.0          N     18   1.144310               864           54.926891
25  1963.0          N     18   1.144310               882           56.071202
26  1998.0          N     17   1.080737               899           57.151939
27  1993.0          N     17   1.080737               916           58.232676
28  2018.0          N     16   1.017165               932           59.249841
29  others     others    641  40.750159              1573          100.000000

`Conclusiones`
- En relación a los valores nulos:
Consideramos que las columnas hace falta un mayor análisis de las columnas que contienen más del 50% de valores nulos, para saber si es reelevante la información que contienen o si en su defecto se podrían agrupar en alguna categoría como 'desconocido' u 'otros' para que en el proceso de limpieza del dataframe no causen problemas.
 